src/main/java/gregtech/api/util/OverlayedFluidHandler.java

```diff
@@ -0,0 +1,188 @@
+    private final OverlayedTank[] originalTanks;
+    private final IMultipleTankHandler overlayed;
+    private boolean allowSameFluidFill = true;
+    private final HashSet<IFluidTankProperties> tankDeniesSameFluidFill = new HashSet<>();
+    private final Map<IMultipleTankHandler, HashSet<FluidKey>> uniqueFluidMap= new HashMap<>();
+
+    public OverlayedFluidHandler(IMultipleTankHandler toOverlay) {
+        this.overlayedTanks = new OverlayedTank[toOverlay.getTankProperties().length];
+        this.originalTanks = new OverlayedTank[toOverlay.getTankProperties().length];
+        this.overlayed = toOverlay;
+    }
+
+    /**
+     * Resets the {slots} array to the state when the handler was
+     * first mirrored
+     */
+
+    public void reset() {
+        for (int i = 0; i < this.originalTanks.length; i++) {
+            if (this.originalTanks[i] != null) {
+                this.overlayedTanks[i] = this.originalTanks[i].copy();
+            }
+        }
+        uniqueFluidMap.forEach((k,v) -> v.clear());
+    }
+
+    public IFluidTankProperties[] getTankProperties() {
+        return overlayed.getTankProperties();
+    }
+
+    private void initTank(int tank) {
+        if (this.overlayedTanks[tank] == null) {
+            IFluidTankProperties fluidTankProperties = overlayed.getTankProperties()[tank];
+            this.originalTanks[tank] = new OverlayedTank(fluidTankProperties);
+            this.overlayedTanks[tank] = new OverlayedTank(fluidTankProperties);
+
+            if (!overlayed.allowSameFluidFill()) {
+                this.allowSameFluidFill = false;
+            }
+            if (overlayed.getTankAt(tank) instanceof NotifiableFluidTankFromList) {
+                NotifiableFluidTankFromList nftfl = (NotifiableFluidTankFromList) overlayed.getTankAt(tank);
+                if (!nftfl.getFluidTankList().get().allowSameFluidFill()) {
+                    this.tankDeniesSameFluidFill.add(overlayed.getTankProperties()[tank]);
+                    uniqueFluidMap.computeIfAbsent(nftfl.getFluidTankList().get(), list -> new HashSet<>());
+                }
+            } else if (!this.allowSameFluidFill) {
+                uniqueFluidMap.computeIfAbsent(overlayed, list -> new HashSet<>());
+            }
+        }
+    }
+
+    public int insertStackedFluidKey(@Nonnull FluidKey toInsert, int amountToInsert) {
+        int insertedAmount = 0;
+        for (int i = 0; i < this.overlayedTanks.length; i++) {
+            initTank(i);
+            // populate the tanks if they are not already populated
+            // if the fluid key matches the tank, insert the fluid
+            OverlayedTank overlayedTank = this.overlayedTanks[i];
+            if (toInsert.equals(overlayedTank.getFluidKey())) {
+                if ((tankDeniesSameFluidFill.contains(overlayed.getTankProperties()[i]) || !this.allowSameFluidFill) ) {
+                    IMultipleTankHandler mth = (IMultipleTankHandler) overlayed;
```

Comment:
This is now a redundant cast after recent changes. Probably could remove the variable and use `overlayed` directly